src/zope/app/apidoc/tests.py

```diff
@@ -13,83 +13,353 @@
+
+        self.assertEqual("((a, b))", getFunctionSignature(locals()['f']))
+
+
+from zope.app.apidoc import static
+
+class TestStatic(unittest.TestCase):
+
+    def _tempdir(self):
+        import tempfile
+        import shutil
+        tmpdir = tempfile.mkdtemp(suffix='.apidoc.TestStatic')
+        self.addCleanup(shutil.rmtree, tmpdir)
+        return tmpdir
+
+    def test_run(self):
+        tmpdir = self._tempdir()
+        static.main(['--max-runtime', '10', os.path.join(tmpdir, 'dir')])
+
+        self.assertIn('static.html',
+                      os.listdir(os.path.join(tmpdir, 'dir', '++apidoc++')))
+
+    def test_run_404(self):
+        tmpdir = self._tempdir()
+        # Fetch a 404 page
+        bad_url = '/++apidoc++/Code/zope/formlib/form/PageEditForm/index.html'
+        maker = static.main(['--max-runtime', '10', # allow for old slow pypy
+                             '--startpage', bad_url,
+                             tmpdir])
+        self.assertEqual(9, maker.counter) # our six default images, plus scripts
+        self.assertEqual(1, maker.linkErrors)
+
+    def test_processLink_errors(self):
+        tmpdir = self._tempdir()
+
+        class ErrorBrowser(object):
+            error_kind = ValueError
+            contents = ''
+            isHtml = False
+            def open(self, url):
+                raise self.error_kind(url)
+
+        class ErrorGenerator(static.StaticAPIDocGenerator):
+            error_kind = ValueError
+
+            def processLink(self, link):
+                b = self.browser
+                self.browser = ErrorBrowser()
+                self.browser.error_kind = self.error_kind
+                super(ErrorGenerator, self).processLink(link)
+                self.browser = b
+
+        maker = static.main(['--max-runtime', '10', os.path.join(tmpdir, 'dir')],
+                            generator=ErrorGenerator)
+        self.assertEqual(7, maker.counter)
+        self.assertEqual(7, maker.linkErrors)
+
+
+        class BadErrorGenerator(ErrorGenerator):
+            error_kind = Exception
+
+        maker = static.main(['--max-runtime', '10', os.path.join(tmpdir, 'dir')],
+                            generator=BadErrorGenerator)
+        self.assertEqual(7, maker.counter)
+        self.assertEqual(0, maker.linkErrors)
+        self.assertEqual(7, maker.otherErrors)
+
+    def test_cleanURL(self):
+        self.assertEqual("http://localhost/",
+                         static.cleanURL("http://localhost/#frogment"))
+
+    def test_completeURL(self):
+        self.assertEqual("http://localhost/index.html",
+                         static.completeURL("http://localhost/"))
+
+    def test_Link_localURL(self):
+
+        self.assertFalse(static.Link('javascript:alert()', '').isLocalURL())
+        self.assertFalse(static.Link('mailto:person@company', '').isLocalURL())
+        self.assertFalse(static.Link("http://external.site/", "http://localhost/").isLocalURL())
```

Comment:
Now I'm curious about 'https://localhost', and 'http://localhost:631' and '//cdn.jquery.com/'.